core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePreferredDynamicFilterTimeout.java

```diff
@@ -0,0 +1,317 @@
+            }
+            throw new IllegalArgumentException("Plan node unsupported " + planNode.getClass().getSimpleName());
+        }
+
+        private static PlanNode getBuildSide(PlanNode planNode)
+        {
+            if (planNode instanceof JoinNode joinNode) {
+                return joinNode.getRight();
+            }
+            else if (planNode instanceof SemiJoinNode semiJoinNode) {
+                return semiJoinNode.getFilteringSource();
+            }
+            else if (planNode instanceof DynamicFilterSourceNode dynamicFilterSourceNode) {
+                return dynamicFilterSourceNode.getSource();
+            }
+            throw new IllegalArgumentException("Plan node unsupported " + planNode.getClass().getSimpleName());
+        }
+
+        private DynamicFilterTimeout getBuildSideState(PlanNode planNode, Symbol dynamicFilterSymbol)
+        {
+            // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+            if (isInputMultiplyingPlanNodePresent(planNode)) {
+                return DynamicFilterTimeout.NO_WAIT;
+            }
+
+            SymbolStatsEstimate symbolStatsEstimate = statsProvider.getStats(planNode).getSymbolStatistics(dynamicFilterSymbol);
+            if (!symbolStatsEstimate.isUnknown() && !isExpandingPlanNodePresent(planNode)) {
+                if (symbolStatsEstimate.getDistinctValuesCount() < awaitedDynamicFilterMaxNdvCount) {
+                    return DynamicFilterTimeout.USE_PREFERRED_TIMEOUT;
+                }
+            }
+
+            double rowCount = getEstimatedMaxOutputRowCount(planNode, statsProvider);
+            if (isNaN(rowCount)) {
+                return DynamicFilterTimeout.UNESTIMATED;
+            }
+            if (rowCount < awaitedDynamicFilterMaxRowCount) {
+                return DynamicFilterTimeout.USE_PREFERRED_TIMEOUT;
+            }
+
+            return DynamicFilterTimeout.NO_WAIT;
+        }
+    }
+
+    private static boolean isInputMultiplyingPlanNodePresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePreferredDynamicFilterTimeout::isInputMultiplyingPlanNode)
+                .matches();
+    }
+
+    private static boolean isInputMultiplyingPlanNode(PlanNode node)
+    {
+        if (node instanceof UnnestNode) {
+            return true;
+        }
+
+        if (node instanceof JoinNode joinNode) {
+            // Skip for cross join
+            if (joinNode.isCrossJoin()) {
+                // If any of the input node is scalar then there's no need to skip cross join
+                return !isAtMostScalar(joinNode.getRight()) && !isAtMostScalar(joinNode.getLeft());
+            }
+
+            // Skip for joins with multi keys since output row count stats estimation can wrong due to
+            // low correlation between multiple join keys.
```

Comment:
it should probably: `can be wrong due to potentially high correlation between join columns, which might lead to underestimation of join output row count`